File: plugin/ChapterHub.Core/ParamAllowlist.cs
using System.Text.Json;

namespace ChapterHub.Core;

/// <summary>
/// packages/contracts/ops/param_allowlist.json as the plugin sees it (SI-4): the ONLY
/// parameters set_parameter may touch, each with the category vocabulary it applies to
/// (walls, doors, windows, furniture, casework, plumbing, electrical; "*" = any). Enrolled
/// beside the MEP catalogs; the gateway checks the same file by target-id prefix before
/// signing and revit-sim by record category — three independent enforcers, one source.
/// </summary>
public sealed class ParamAllowlist
{
    public sealed record Entry(string Name, string Kind, IReadOnlyList<string> Categories);

    public static readonly IReadOnlySet<string> Vocabulary = new HashSet<string>
    {
        "walls", "doors", "windows", "furniture", "casework", "plumbing", "electrical",
    };

    private readonly IReadOnlyDictionary<string, Entry> _byName;

    public string AllowlistVersion { get; }
    public IReadOnlyList<Entry> Entries { get; }

    private ParamAllowlist(string version, IReadOnlyList<Entry> entries)
    {
        AllowlistVersion = version;
        Entries = entries;
        _byName = entries.ToDictionary(e => e.Name);
    }

    public static ParamAllowlist FromJson(string json)
    {
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        var entries = root.GetProperty("params").EnumerateArray().Select(p =>
        {
            var categories = p.GetProperty("categories").EnumerateArray().Select(c => c.GetString()!).ToList();
            foreach (var c in categories)
                if (c != "*" && !Vocabulary.Contains(c))
                    throw new JsonException($"param_allowlist: unknown category {c}");
            return new Entry(p.GetProperty("name").GetString()!, p.GetProperty("kind").GetString()!, categories);
        }).ToList();
        return new ParamAllowlist(root.GetProperty("allowlist_version").GetString()!, entries);
    }

    public bool Contains(string param) => _byName.ContainsKey(param);

    public string? Kind(string param) => _byName.TryGetValue(param, out var e) ? e.Kind : null;

    /// <summary>Allowed when the param is listed and either carries "*" or names the target's
    /// category; a target with no vocabulary category (null) may only take "*" params.</summary>
    public bool IsAllowed(string param, string? category)
    {
        if (!_byName.TryGetValue(param, out var entry)) return false;
        if (entry.Categories.Contains("*")) return true;
        return category is not null && entry.Categories.Contains(category);
    }

    /// <summary>Finish/product/spec/comment params carry text: the value the executor writes
    /// must be a string (revit-sim's param_type_mismatch rule).</summary>
    public bool RequiresString(string param) =>
        Kind(param) is "finish" or "product" or "spec" or "comment";
}

File: services/gateway/src/envelope/param-allowlist.ts
// SI-4 at the signer (Phase 7, P7-10): every set_parameter op must name a parameter in
// ops/param_allowlist.json that allows the target's category. The category is derived from
// the logical-id prefix (the layout id patterns: W walls, D doors, N windows, K casework,
// F furniture, E electrical); anything else has no category and may only take "*" params.
// F -> furniture is sound because every allowlisted param that lists `plumbing` also lists
// `furniture` (param-allowlist.test.ts pins that invariant, so an allowlist edit that breaks
// it forces a hookups-aware rule deliberately). The sim re-checks by record category and
// the plugin by BuiltInCategory + storage type — three independent enforcers.
import { paramAllowlist } from "@chapter/contracts";
import type { OpInput } from "./builder.js";

export type ParamCategory =
  | "walls"
  | "doors"
  | "windows"
  | "furniture"
  | "casework"
  | "plumbing"
  | "electrical";

const PREFIX_CATEGORY: Record<string, ParamCategory> = {
  W: "walls",
  D: "doors",
  N: "windows",
  K: "casework",
  F: "furniture",
  E: "electrical",
};

const BY_NAME = new Map(paramAllowlist.params.map((p) => [p.name, p]));

export function categoryForTarget(targetId: string): ParamCategory | null {
  const m = /^([A-Z]{1,2})-\d{2,4}$/.exec(targetId);
  return m ? (PREFIX_CATEGORY[m[1]!] ?? null) : null;
}

export interface AllowlistProblem {
  index: number;
  op: string;
  reason: "param_not_allowlisted";
  detail: string;
}

export function isParamAllowed(param: string, category: ParamCategory | null): boolean {
  const entry = BY_NAME.get(param);
  if (!entry) return false;
  if (entry.categories.includes("*")) return true;
  return category !== null && entry.categories.includes(category);
}

/** null when every set_parameter op is allowlisted for its target's category. */
export function checkParamAllowlist(ops: OpInput[]): AllowlistProblem | null {
  for (let index = 0; index < ops.length; index++) {
    const op = ops[index]!;
    if (op.op !== "set_parameter") continue;
    const param = String(op.args["param"] ?? "");
    const target = String(op.args["target_id"] ?? "");
    const entry = BY_NAME.get(param);
    if (!entry) {
      return { index, op: op.op, reason: "param_not_allowlisted", detail: `param ${param} is not in ops/param_allowlist.json` };
    }
    if (entry.categories.includes("*")) continue;
    const category = categoryForTarget(target);
    if (category === null) {
      return {
        index, op: op.op, reason: "param_not_allowlisted",
        detail: `target ${target} has no allowlisted category; ${param} allows ${entry.categories.join(", ")}`,
      };
    }
    if (!entry.categories.includes(category)) {
      return {
        index, op: op.op, reason: "param_not_allowlisted",
        detail: `${param} is not allowed on ${category} (${target}); allowed: ${entry.categories.join(", ")}`,
      };
    }
  }
  return null;
}
